File: array_and_strings/sliding_window/c++/product_less_than_k/product_less_than_k.cpp
// product_less_than_k.cpp
#include "product_less_than_k.hpp"
#include <vector>

int numSubarraysProductLessThanK(const std::vector<int> &nums, const int &k)
{
    // If k <= 1, then there is no valid window
    if (k <= 1)
    {
        return 0;
    }

    // Initialization
    int currProduct = 1, left = 0, numSubarrays = 0;

    for (int right = 0; right < nums.size(); right++)
    {
        // Product of all elements in the current window
        currProduct *= nums[right];

        // Check if window becomes invalid, i.e. currProduct exceeds or equals to k
        while (currProduct >= k)
        {
            // Divide the product by the element the left pointer points to
            currProduct /= nums[left];
            // Slide the window towards the right
            left++;
        }

        // Add the total number of subarrays in the current window to the total
        // This is because given the current window (left, right), there are (right - left + 1) subarrays
        numSubarrays += right - left + 1;
    }

    return numSubarrays;
}

File: array_and_strings/sliding_window/python3/product_less_than_k.py
from typing import List


def num_subarrays_product_less_than_k(nums: List[int], k: int) -> int:
    """
    Given an array of positive integers nums and an integer k, return the number of subarrays where the product of all the elements in the subarray is strictly less than k.

    Parameters
    ----------
    nums : List[int]
        A list of positive integers
    k : int
        An integer

    Returns
    -------
    int
        The number of contiguous subarrays where the product of all the elements in the subarray is strictly less than k.
    """
    # If k <= 1, there is no valid subarray where the product of its elements are strictly <= 1
    if k <= 1:
        return 0

    left = num_subarrays = 0
    curr_product = 1

    for right in range(len(nums)):
        # Current product of all elements in the current window
        curr_product *= nums[right]
        # The window becomes invalid once the product is no longer strictly less than k
        while curr_product >= k:
            # Divide the current product by the element the left pointer points to, using // integer division
            curr_product //= nums[left]
            # Slide the window towards the right
            left += 1

        # Instead of updating, we add the current window size to the total number of subarrays variable
        # This is because given the current window (left, right), there are (right - left + 1) subarrays
        num_subarrays += right - left + 1

    return num_subarrays


def main() -> int:
    nums = [10, 5, 2, 6, 7]
    k = 35

    num_subarrays = num_subarrays_product_less_than_k(nums, k)

    print(
        f"The input {nums} with k = {k} contains {num_subarrays} subarrays where the product of all its elements is strictly less than k"
    )

    return 0


if __name__ == "__main__":
    main()
